langchain/prompts/prompt.py

```diff
@@ -1,53 +1,98 @@
 """Prompt schema definition."""
 from __future__ import annotations
 
+import re
+from abc import ABC, abstractmethod
 from pathlib import Path
 from string import Formatter
-from typing import Any, Dict, List, Union
+from typing import Any, Dict, List, Optional, Union
 
 from pydantic import BaseModel, Extra, root_validator
 
 from langchain.prompts.base import (
     DEFAULT_FORMATTER_MAPPING,
     BasePromptTemplate,
+    PromptValue,
+    StringPromptValue,
     check_valid_template,
 )
 
 
-class PromptTemplate(BasePromptTemplate, BaseModel):
-    """Schema to represent a prompt for an LLM.
+class BaseOutputParser(BaseModel, ABC):
+    """Class to parse the output of an LLM call."""
 
-    Example:
-        .. code-block:: python
+    @abstractmethod
+    def parse(self, text: str) -> Union[str, List[str], Dict[str, str]]:
+        """Parse the output of an LLM call."""
 
-            from langchain import PromptTemplate
-            prompt = PromptTemplate(input_variables=["foo"], template="Say {foo}")
-    """
+    @property
+    def _type(self) -> str:
+        """Return the type key."""
+        raise NotImplementedError
 
-    input_variables: List[str]
-    """A list of the names of the variables the prompt template expects."""
+    def dict(self, **kwargs: Any) -> Dict:
+        """Return dictionary representation of output parser."""
+        output_parser_dict = super().dict()
+        output_parser_dict["_type"] = self._type
+        return output_parser_dict
 
-    template: str
+
+class ListOutputParser(BaseOutputParser):
```

Comment:
Conversational bots tend to include extra verbiage "Here you go: XXX" or "XXX. That does what you want, but lemme know if you want to change it". And so you can ask the chatbot to include it in markdown/xml, and it might do "Here you go: <result>A,B,C</result>". So can perhaps be done with RegexParser, but may want to feed RegexParser into CommaSeparatedListOutput when dealing with conversational responses. Maybe it's not worth chaining outputparsers here, but sharing one of the differences for chatbots.

Note that proper few-shot-prompting can solve this and make the chatbots less chatty, so maybe this isn't as important if done well.

Reply:
Yeah the output parsers are mostly intended for instruct-style llm output, unclear how useful they'll be for chat output